### PR TITLE
WIP: add tap/hold actions to entity rows

### LINKED ISSUE #238: Entity type rows should support tap_action
*@SeanPM5 commented on Feb 17, 2019, 5:45 AM UTC:*

**Home Assistant release with the issue:**

0.87.1

**Description of problem:**

Lovelace entity rows do not support tap_action or hold_action according to the [docs](https://www.home-assistant.io/lovelace/entities/), I feel they should. The feature set should be consistent where it makes sense.

Here is an example use case. This image is from my default_view showing a few key statistics, and I want to link the "Time Driving Today" row to my Cars view.

[Lovelace feature request](https://camo.githubusercontent.com/277e80d31c2d7441dd7ce1138777acc80bebd66d/68747470733a2f2f692e696d6775722e636f6d2f4470506f7930582e706e67)

Example config:

```
        - type: entities
          entities:
            - sensor.time_away_today
            - entity: sensor.time_driving_today
              name: Time Driving Today
              tap_action:
                action: navigate
                navigation_path: /lovelace/cars/

```

Instead of navigating it simply opens the more-info popup. This same code works in a glance card, but not an entities type card.

*This issue was moved by [iantrich](https://github.com/iantrich) from [home-assistant/home-assistant-polymer#2774](https://github.com/home-assistant/home-assistant-polymer/issues/2774).*

### PR DESCRIPTION
**Pull request by iantrich**
Fixes https://github.com/home-assistant/ui-schema/issues/238

### CODE REVIEW

**Comment by iantrich** [comment]
ready for review

**Comment by klaasnicolaas** [comment]
https://github.com/home-assistant/home-assistant.io/pull/8705

File: src/panels/lovelace/cards/hui-entities-card.ts [hunk 1]
```diff
@@ -168,29 +172,24 @@ class HuiEntitiesCard extends LitElement implements LovelaceCard {
     if (this._hass) {
       element.hass = this._hass;
     }
+
     if (
-      entityConf.entity &&
-      !DOMAINS_HIDE_MORE_INFO.includes(computeDomain(entityConf.entity))
+      ((!entityConf.tap_action ||
+        entityConf.tap_action.action === "more-info") &&
+        !DOMAINS_HIDE_MORE_INFO.includes(computeDomain(entityConf.entity))) ||
```
Comment (balloob) [inline src/panels/lovelace/cards/hui-entities-card.ts]: We shouldn't be smart here. If a user picks `more-info`, we are going to add `more-info`. Only use the `DOMAINS_HIDE_MORE_INFO` if a user did not specify a tap action.

File: src/panels/lovelace/components/hui-generic-entity-row.ts [hunk 2]
```diff
@@ -43,32 +45,39 @@ class HuiGenericEntityRow extends LitElement {
...
-              ? html`
-                  <ha-relative-time
-                    .hass="${this.hass}"
-                    .datetime="${stateObj.last_changed}"
-                  ></ha-relative-time>
-                `
-              : ""}
+      <div
+        class="host"
+        @ha-click="${this._handleTap}"
```
Comment (balloob) [inline src/panels/lovelace/components/hui-generic-entity-row.ts]: I don't think that this is a good idea.

- How does this work with interactions with the state controls like turning a light on/off ? Would that still be able to do long press too ?
- What about entity rows that do not use generic-entity-row? For example, if we add the mini-media-player, this won't work?

File: src/panels/lovelace/entity-rows/hui-input-select-entity-row.ts [hunk 3]
```diff
@@ -51,25 +54,26 @@ class HuiInputSelectEntityRow extends LitElement implements EntityRow {
...
-      <state-badge .stateObj="${stateObj}"></state-badge>
-      <paper-dropdown-menu
-        selected-item-label="${stateObj.state}"
-        @selected-item-label-changed="${this._selectedChanged}"
-        label="${this._config.name || computeStateName(stateObj)}"
-      >
-        <paper-listbox
-          slot="dropdown-content"
-          selected="${stateObj.attributes.options.indexOf(stateObj.state)}"
+      <hui-generic-entity-row .hass="${this.hass}" .config="${this._config}">
```
Comment (balloob) [inline src/panels/lovelace/entity-rows/hui-input-select-entity-row.ts]: This is not the same. Now it will render the name besides the dropdown. This was not done before because the dropdown already had the name set as a label.

File: src/panels/lovelace/entity-rows/hui-text-entity-row.ts [hunk 4]
```diff
@@ -5,15 +5,18 @@ import {
   property,
   CSSResult,
   css,
+  customElement,
```
Comment (balloob) [inline src/panels/lovelace/entity-rows/hui-text-entity-row.ts]: Can you split out all these changes into a separate PR that we can merge?

Comment (iantrich) [reply src/panels/lovelace/entity-rows/hui-text-entity-row.ts]: sure

File: src/panels/lovelace/entity-rows/hui-input-select-entity-row.ts (hunk 3 above)
Comment (iantrich) [reply src/panels/lovelace/entity-rows/hui-input-select-entity-row.ts]: lame...lol

File: src/panels/lovelace/components/hui-generic-entity-row.ts (hunk 2 above)
Comment (iantrich) [reply src/panels/lovelace/components/hui-generic-entity-row.ts]: * in my tests I didn't have issues when using toggle rows. are you asking if you can long press the toggle?
* no, it would not. so you would prefer I add click/hold to each row instead of the base?

Comment (iantrich) [reply src/panels/lovelace/components/hui-generic-entity-row.ts]: @balloob any further input on what you think should happen here?

File: src/panels/lovelace/cards/hui-entities-card.ts [hunk 5]
```diff
@@ -4,21 +4,22 @@ import {
   PropertyDeclarations,
   PropertyValues,
   TemplateResult,
+  customElement,
```
Comment (iantrich) [inline src/panels/lovelace/cards/hui-entities-card.ts]: I rebased on dev, not sure why it is still showing this as if it is a new change...?

Comment (balloob) [reply src/panels/lovelace/cards/hui-entities-card.ts]: When you rebase, make sure you rebase off the latest version of upstream

```
git fetch upstream
git rebase upstream/dev
```

File: src/panels/lovelace/cards/hui-entities-card.ts [hunk 6]
```diff
@@ -168,29 +172,22 @@ class HuiEntitiesCard extends LitElement implements LovelaceCard {
     if (this._hass) {
       element.hass = this._hass;
     }
+
     if (
-      entityConf.entity &&
-      !DOMAINS_HIDE_MORE_INFO.includes(computeDomain(entityConf.entity))
```
Comment (balloob) [inline src/panels/lovelace/cards/hui-entities-card.ts]: Why did we remove this from the default?

File: src/panels/lovelace/cards/hui-entities-card.ts [hunk 7]
```diff
@@ -168,29 +172,22 @@ class HuiEntitiesCard extends LitElement implements LovelaceCard {
...
+      !entityConf.tap_action ||
+      (entityConf.tap_action && entityConf.tap_action.action !== "none")
     ) {
       element.classList.add("state-card-dialog");
```
Comment (balloob) [inline src/panels/lovelace/cards/hui-entities-card.ts]: Why do we only add this if there is a default action? It feels weird that this element should even know about any clicking being added, should be responsibility of generic-entity-row ?

Comment (iantrich) [reply src/panels/lovelace/cards/hui-entities-card.ts]: Not all have generic-entity-row, input-select-entity-row in particular does not use generic-entity-row

Comment (balloob) [reply src/panels/lovelace/cards/hui-entities-card.ts]: Actually, I wonder if the longpress directive should add this?

File: src/panels/lovelace/cards/hui-entities-card.ts (hunk 5 above)
Comment (iantrich) [reply src/panels/lovelace/cards/hui-entities-card.ts]: That is what I always do, perhaps I botched it

File: src/panels/lovelace/cards/hui-entities-card.ts (hunk 7 above)
Comment (balloob) [reply src/panels/lovelace/cards/hui-entities-card.ts]: hm, no that feels weird, I guess the component that renders a longpress directive should take care of it

**Comment by balloob** [comment]
I'm starting to have second thoughts about this. Allowing people to override the name part of an entity row… are we really improving things? What is the use case ? Discoverability is poor of what an action will do, so we will only allow people to build more confusing interfaces.

**Comment by iantrich** [comment]
If you believe that, then you the argument would hold for ALL our tap/hold actions which is most of our cards and elements; allowing it for rows makes them conform to what is common.

e.g. If I want clicking on my climate row to start a script to toggle it for away mode and long-press for more-info, I think that is a valid and useful action. Not allowing it would make it so I have to have a script row and a climate row to achieve the same thing.

In the end, it should be up to me how I want things to be handled as a user.

**Comment by balloob** [comment]
We introduced click/long press for the picture elements. There they make sense. 

I think that this is a feature creep that will make editors and features more complex. If we don't show the away state on a climate card, how will you know the call was successful? There is no toast anymore.

**Comment by iantrich** [comment]
How does it make sense for only picture-elements? I strongly disagree your arguments. If you're concerned about not knowing if a service call was successful, we should have not taken away toasts, because that argument touches anything that does implement actions already; elements, picture-entity, picture-glance, entity-button, and glance.

This is not feature creep, in my opinion. It is an alignment with the other items I listed above.

**Comment by SeanPM5** [comment]
Tap and hold actions are already supported in the non picture cards like Glance and Entity Button for example. Here is an example of one of my system glances which i find quite useful.

```
      - type: glance
        columns: 4
        show_name: true
        entities:
          - entity: sensor.home_assistant_v2_db
            name: Database
            icon: mdi:database
            hold_action:
              action: call-service
              service: recorder.purge # purges db keeping last 3 days only
              data:
                keep_days: 3
                repack: true
		  
          - entity: sensor.home_assistant_log
            name: Logfile
            icon: mdi:file-alert-outline
            tap_action:
              action: navigate
              navigation_path: /dev-info # goes to log in HA when tapped
            hold_action:
              action: call-service
              service: system_log.clear # clears log in HA when long-pressed
		  
          - entity: sensor.time_online
            name: Uptime
            hold_action:
              action: call-service
              service: homeassistant.restart # restarts HA when long-pressed
		  
          - entity: sensor.installed_version
            tap_action:
              action: navigate
              navigation_path: /hassio/dashboard # goes to Hassio update section when tapped
```

It makes no sense to me that I can do this stuff in a glance card, but not in an entities card... The only difference to me (as an end user) is that one displays stuff horizontally and the other displays vertically. 

I think it's a good idea to have consistency amongst the core cards where it makes sense.

**Comment by iantrich** [comment]
Some reasoning from the past 
https://github.com/home-assistant/ui-schema/issues/189

**Comment by balloob** [comment]
There is a big difference between glance and entities card, it's not just vertical vs horizontal. Entities card has entity rows and entity rows have controls. And because there are no controls, the icons look clickable. 

For entities card, it's a name next to actual controls. What we're proposing here is a UX footgun: the label next to the controls can now be configured to do anything. We should empower people to build UIs that everyone in their house can use. We have kept a tight lit on features that would be confusing for users. Shipping less features, but each feature is easier to understand and results in easier to understand UIs, is a big plus. Too many projects are getting feature creep'd to hell and get replaced. 

The Lovelace built-in cards, entity rows and elements should not aim to cover every feature or aim to please everyone. The built-ins should focus on creating an easy to use set of elements to build UIs that cover 80% of the use cases.

**Comment by iantrich** [comment]
Icons are clickable in our UI, that is a fact, even in the entities card; we're merely giving it more control. And I still argue this is not creep; this is alignment with our other cards.

**Comment by balloob** [comment]
Alignment only makes sense if the feature actually suits the card. I argue that in this case, UX-wise it doesn't make sense.

Can you give me some examples of how this feature will be used that also gives a good user experience? The reason I didn't like the "away mode" toggle example you gave earlier was because we don't show the current away mode on the controls, so allowing it to be toggled, but you don't know what you're toggling to, is bad UX. If away mode is shown on the card, maybe we should allow changing it from the controls too? Custom rows like the mini media player show that we still have a lot of room to expand controls.

**Comment by SeanPM5** [comment]
Here is my above Glance example with the same exact thing re-created in an Entities card. 

image

And another example of a room presence card in both Glance and Entities format:

image

In my opinion, the Glance cards do not appear to be any more clickable than the Entity cards. There's no obvious visual cues like button shapes or different colored / underlined text to indicate they're clickable. 

The only reason I know that the Glances in theses screenshots have associated click actions is because I added the necessary tap_action and hold_action commands to them myself. And that same reasoning would hold true for the Entities card too.

It makes sense to keep the default behavior as simple as possible, but this is not default behavior. This is an option that the user must first manually add and configure themselves. And when you're the one configuring something to your liking, you would know _exactly_ how it works. There shouldn't be any confusion imo.

**Comment by iantrich** [comment]
@SeanPM5 you're input on a compromise:
image
A state/action generic row
You specify the state to view and the action that may should be taken. I'm thinking that call-service and navigate would be the only options for the action.

**Comment by iantrich** [comment]
I'm thinking the state would be best displayed below the name with a smaller size and the secondary color

**Comment by SeanPM5** [comment]
For me personally, much of the appeal here was that it was minimal, so I'm not a huge fan of the "Go" button idea. It takes up room from and sorta draws attention away from the entity name and state. And I assume this would likely mean that only the button is tappable now, rather than the entire row?

I don't know if this is possible, but here is how Apple kind of handles this scenario with iOS:

image

The icons on the right indicate it takes you to another screen, but the entire row is still tappable. So maybe you can put a similar icon there (mdi:chevron-right) instead of the button? I think that would be a slightly better compromise if possible.

**Comment by iantrich** [comment]
I do like that design much more

**Comment by iantrich** [comment]
https://material.io/design/components/lists.html#anatomy

**Comment by iantrich** [comment]
Thinking about resurrecting this and do believe that @SeanPM5 example of glance vs entities highlights that this is a matter of alignment

**Comment by iantrich** [comment]
@balloob what do you think about adding just hold_action to rows, maintaining our current default tap actions?

**Comment by SeanPM5** [comment]
I do still believe this would be useful, even more so now that Home Assistant Cast is a thing. 

Cast relies on functionality like this in order to navigate between views on device. A lot of other card types support this, and even the weblink special row (which can be placed _inside_ of an entities card) supports this now. Feels a little weird that entity row is the one place it's omitted IMO.

**Comment by iantrich** [comment]
Cast is a very good point. In fact makes the argument for tap action to be included
